crates/sui/tests/shared_objects_version_tests.rs

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::path::PathBuf;
+use sui_config::{NetworkConfig, ValidatorInfo};
+use sui_node::SuiNode;
+use sui_types::base_types::{ObjectRef, SequenceNumber};
+use sui_types::error::SuiResult;
+use sui_types::messages::{CallArg, ObjectArg, TransactionEffects};
+use sui_types::object::{Object, Owner, OBJECT_START_VERSION};
+use test_utils::authority::{spawn_test_authorities, test_authority_configs};
+use test_utils::messages::move_transaction;
+use test_utils::objects::test_gas_objects;
+use test_utils::transaction::{
+    publish_package, submit_shared_object_transaction, submit_single_owner_transaction,
+};
+
+#[tokio::test]
+async fn fresh_shared_objects_get_start_version() {
+    let mut env = TestEnvironment::new().await;
+    let (_, owner) = env.create_shared_counter().await;
+    assert!(is_shared_at(&owner, OBJECT_START_VERSION));
+}
+
+#[tokio::test]
+async fn objects_transitioning_to_shared_remember_their_previous_version() {
+    let mut env = TestEnvironment::new().await;
+    let (counter, _) = env.create_counter().await;
+
+    let (counter, _) = env.increment_owned_counter(counter).await;
+    assert_ne!(counter.1, OBJECT_START_VERSION);
+
+    let (counter, owner) = env.share_counter(counter).await;
```

Comment:
We should use the shared object again to make sure the owner doesn't change?

Reply:
👍

Reply:
Actually, I realised that the test function below checks this.